ingestion/src/metadata/ingestion/source/database/database_source.py

```diff
@@ -0,0 +1,319 @@
+    Column,
+    DataModel,
+    Table,
+    TableData,
+    TableProfile,
+)
+from metadata.generated.schema.type.entityReference import EntityReference
+from metadata.ingestion.api.common import Entity
+from metadata.ingestion.api.source import Source
+from metadata.ingestion.models.ometa_table_db import OMetaDatabaseAndTable
+from metadata.ingestion.models.ometa_tag_category import OMetaTagAndCategory
+from metadata.ingestion.models.table_metadata import DeleteTable
+from metadata.ingestion.ometa.utils import ometa_logger
+from metadata.orm_profiler.orm.converter import ometa_to_orm
+from metadata.orm_profiler.profiler.default import DefaultProfiler
+from metadata.utils.filters import filter_by_schema, filter_by_table
+from metadata.utils.fqdn_generator import get_fqdn
+
+logger = ometa_logger()
+
+
+@dataclass  # type: ignore[misc]
+class DatabaseSourceService(Source, metaclass=ABCMeta):
+    @abstractmethod
+    def get_databases(self) -> Iterable[Inspector]:
+        """
+        Method Yields Inspector objects for each aviable database
+        """
+        pass
+
+    @abstractmethod
+    def get_schemas(self, inspector: Inspector) -> str:
+        """
+        Method Yields schemas aviable in database
+        """
+        pass
+
+    @abstractmethod
+    def standardize_schema_table_names(
+        self, schema: str, table: str
+    ) -> Tuple[str, str]:
+        """
+        Method formats Table & Schema names if required
+        """
+        pass
+
+    @abstractmethod
+    def _get_table_description(
+        self, schema: str, table_name: str, table_type: str, inspector: Inspector
+    ) -> str:
+        """
+        Method returns the table level comment
+        """
+        pass
+
+    @abstractmethod
+    def _is_partition(self, table_name: str, schema: str, inspector: Inspector) -> bool:
+        """
+        Method to check if the table is partitioned table
+        """
+        pass
+
+    @abstractmethod
+    def _get_data_model(self, database: str, schema: str, table_name: str) -> DataModel:
+        pass
+
+    @abstractmethod
+    def fetch_sample_data(self, schema: str, table_name: str) -> Optional[TableData]:
+        """
+        Method to fetch sample data form table
+        """
+        pass
+
+    @abstractmethod
+    def get_table_names(self, schema: str, inspector: Inspector) -> Optional[List[str]]:
+        """
+        Method to fetch table & view names
+        """
+        pass
+
+    @abstractmethod
+    def _get_columns(
+        self, schema: str, table_name: str, inspector: Inspector
+    ) -> Optional[List[Column]]:
+        """
+        Method to fetch table columns data
+        """
+        pass
+
+    @abstractmethod
+    def get_view_definition(table_type, table_name, schema, inspector) -> Optional[str]:
+        """
+        Method to fetch view definition
+        """
+        pass
+
+    def _get_database(self, database: Optional[str]) -> Database:
```

Comment:
This is what we were discussing the other day. Here we have `database` as an argument which as an `Optional[str]`. In the function below, database is a `Database`. We should try to assign names that describe what is going on and stick to a single convention. If `database` is `Database`, then for the names let's use `database_name`. This is just an example, any other naming works, but we need to follow that across the whole code!